java-squid/src/main/java/org/sonar/java/se/ExplodedGraph.java

```diff
@@ -46,33 +51,196 @@ Node getNode(ProgramPoint programPoint, @Nullable ProgramState programState) {
     return result;
   }
 
-  public static class ProgramPoint {
+  public static interface ProgramPoint {
+
+    Block block();
+
+    ProgramPoint next();
+
+    Tree terminator();
+
+    boolean atEnd();
+
+    Tree currentElement();
+
+    boolean hasBlock();
+
+    Collection<Block> successors();
+
+  }
+
+  public static class FinallyProgramPoint implements ProgramPoint {
+
     private int hashcode;
-    final CFG.Block block;
-    final int i;
+    private final List<CFG.Block> blocks;
+    private final int blockStep;
+    private final int step;
+
+    public FinallyProgramPoint(List<CFG.Block> blocks) {
+      this(blocks, 0, 0);
+    }
+
+    private FinallyProgramPoint(List<CFG.Block> blocks, int blockStep, int step) {
+      this.blocks = new ArrayList<>(blocks);
+      this.step = step;
+      this.blockStep = adjustBlockIndex(blockStep);
+    }
+
+    private int adjustBlockIndex(int blockStep) {
+      int n = blockStep;
+      while (blocks.get(n).elements().isEmpty()) {
+        n += 1;
+        if (n == blocks.size()) {
+          break;
+        }
+      }
+      return n;
+    }
+
+    @Override
+    public int hashCode() {
+      if (hashcode == 0) {
+        long hash = step;
+        for (Block block : blocks) {
+          hash = block.id() * 31 + hash >> 16;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)